### PR TITLE
`Adaptive learning`: Fix competency progress update for exercise and lecture unit operations

### PR DESCRIPTION
**Pull request by JohannesStoehr**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context

The competency progress was not updated correctly for different exercise and lecture operations. This could lead to wrong information shown to the student as well as wrong suggestions on the learning path.

### Description

Each operation that an instructor can do regarding exercises and lecture units should now correctly update the competency progress. These operations include:
- Creating an exercise/lecture unit and directly link it to a competency
- Updating an exercise/lecture unit and change the linked competencies
- Deleting an exercise/lecture unit that is linked to a competency
- Deleting a lecture that contains lecture units that are linked to competencies
- Import an exercise and directly link it to a competency
IMPORTANT: The corresponding updates are done asynchronously, so it might take a short while until the changes are noticeable.

Any operation that changes the student score (e.g. assessment or changing programming test cases) already lead to a reevaluation of competency progresses. But this is only checked every minute and then asynchronously updated, so it might take a while...

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student with some progress in a competency

1. Log in to Artemis
2. Do different operations as an instructor and check that the competency progress increases/decreases for the student accordingly

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| AttachmentUnitService.java | 89% | ✅ |
| ExerciseDeletionService.java | 96% | ✅ |
| ExerciseImportService.java | 97% | ✅ |
| FileUploadExerciseImportService.java | 100% | ✅ |
| LectureService.java | 98% | ✅ |
| LectureUnitService.java | 88% | ✅ |
| ModelingExerciseImportService.java | 96% | ✅ |
| ParticipationService.java | 90% | ✅ |
| TextExerciseImportService.java | 100% | ✅ |
| CompetencyProgressService.java | 95% | ✅ |
| ProgrammingExerciseService.java | 87% | ✅ |
| QuizExerciseImportService.java | 94% | ✅ |
| CompetencyResource.java | 94% | ✅ |
| FileUploadExerciseResource.java | 100% | ✅ |
| LectureResource.java | 90% | ✅ |
| ModelingExerciseResource.java | 96% | ✅ |
| QuizExerciseResource.java | 92% | ✅ |
| TextExerciseResource.java | 96% | ✅ |
| AttachmentUnitResource.java | 92% | ✅ |
| OnlineUnitResource.java | 87% | ✅ |
| TextUnitResource.java | 96% | ✅ |
| VideoUnitResource.java | 90% | ✅ |
| ProgrammingExerciseExportImportResource.java | 91% | ✅ |
| ProgrammingExerciseResource.java | 81% | ✅ |

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved entity disconnection to maintain database integrity.

- **New Features**
  - Enhanced exercise repositories to fetch entities with related competencies.
  - Added methods to retrieve competencies and user scores for various exercises.
  - Added functionality to fetch users within a specific course.
  - Introduced competency progress tracking for updated learning objects.

- **Improvements**
  - Refined methods to ensure eager loading of competencies in exercises.
  - Enhanced dependencies and functionality in services to handle competencies.

### CODE REVIEW

**Review by BBesrour — CHANGES_REQUESTED** [review]
Tested on TS3

The competency isn't updated in this scenario:
1. As student, mark a lecture unit as completed
2. As Instructor, create (or import) a new competency and link it to the completed competency.
3. Student competency not updated.

image
-
image
-
image

File: src/main/java/de/tum/in/www1/artemis/service/ExerciseImportService.java
Comment (pzdr7) [inline src/main/java/de/tum/in/www1/artemis/service/ExerciseImportService.java:65]: What is supposed to happen when I import an exercise from another course? e.g. I imported an exercise from a different course that had two linked competencies. When viewing the exercise in my course, the linked competencies were still there, even though they are inaccessible in my course.

Then, when I tried to delete the other course, I got an internal server error because the competencies are still referenced in the other course:

    "detail": "could not execute batch [Batch entry 0 delete from competency where id=('97'::int8) was aborted: ERROR: update or delete on table \"competency\" violates foreign key constraint \"FKsqsxg20yruu16ca9x383wnhci\" on table \"competency_exercise\"\n  Detail: Key (id)=(97) is still referenced from table \"competency_exercise\".  Call getNextException to see other errors in the batch.] [delete from competency where id=?]; SQL [delete from competency where id=?]; constraint [FKsqsxg20yruu16ca9x383wnhci]",
    "path": "/api/admin/courses/74",

**Review by pzdr7 — CHANGES_REQUESTED** [review]
Tested on TS3. See my comment below

File: src/main/java/de/tum/in/www1/artemis/service/ExerciseImportService.java
Comment (JohannesStoehr) [reply src/main/java/de/tum/in/www1/artemis/service/ExerciseImportService.java:65]: They should not be imported. Fixed!

File: src/main/webapp/app/entities/quiz/quiz-exercise.model.ts
Comment (JohannesStoehr) [reply src/main/webapp/app/entities/quiz/quiz-exercise.model.ts]: Good catch! Fixed

**Review by BBesrour — DISMISSED** [review]
Thanks for addressing the issue!
Tested on TS3, works as described

**Review by Jan-Thurner — COMMENTED** [review]
I am not sure if I understand the testing steps correctly. I have a programming exercise linked to a competency. No matter how high or low the mastery threshold is defined, the mastery progress remains the same.
image

For a mastery with multiple lectures, the mastery threshold does not seem to influence the mastery progress at all. It is always either 0%, 50% or 100%. 
image

**Review by BBesrour — APPROVED** [review]

**Review by pzdr7 — COMMENTED** [review]
Tested on TS3. The updates work as expected, but I did find one potential edge case (see the video below). Rapid changes to the units could result in the wrong progress being shown.

Is this relevant in practice?

https://github.com/ls1intum/Artemis/assets/38403547/b7b4d8bb-107e-43e3-bedb-723e684727e1

**Comment by JohannesStoehr** [comment]
> Tested on TS3. The updates work as expected, but I did find one potential edge case (see the video below). Rapid changes to the units could result in the wrong progress being shown.

This requires a deeper change since we currently block any update when the student progress was already updated in the last second. So I would consider this as out of scope for this PR

**Review by pzdr7 — APPROVED** [review]
Approving since the bug I found was out of scope for this PR :+1:

**Review by Jan-Thurner — APPROVED** [review]
Tested on TS3, works as described.

**Review by florian-glombik — COMMENTED** [review]
The Code looks good to me, I left a few suggestions that I think would be beneficial for future maintenance

**Review by Jan-Thurner — DISMISSED** [review]

**Review by JohannesWt — DISMISSED** [review]
Code 👍

**Review by pzdr7 — DISMISSED** [review]
Reapprove after refactoring

**Review by florian-glombik — DISMISSED** [review]
Approve Code, thanks for the adjustments!

**Review by MaximilianAnzinger — DISMISSED** [review]
Maintainer approved

**Review by florian-glombik — DISMISSED** [review]
Re-approve code after merge conflict https://github.com/ls1intum/Artemis/pull/8976/commits/fb204cd974bef888ee6d54512d1b538995529ce8

**Review by BBesrour — APPROVED** [review]

**Review by egekurt123 — DISMISSED** [review]
reapprove after solving merge conflicts

**Review by az108 — APPROVED** [review]
Tested on TS1 and works as expected. And Code LGTM aswell after merge changes.
Before change:
image
After change:
image

**Pull request merged**
